### PR TITLE
shopping cart visitor

### PR DESCRIPTION
**Pull request by TalKoren11**
Done for now, I will need to change it once I done with payment and shipping classes.
And there is still room for more functionality to be added such as if the item is out of stock immediately delete it from cart but that is for later.

### CODE REVIEW

File: src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package Domain.Exceptions;
```
Comment (inbarbenchaim) [inline src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java:1]: maybe in external folder- cause also needed in the frontend and the backend?
if not- leave it here.

File: src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+public class ItemOutOfStockExepction extends RuntimeException {
```
Comment (inbarbenchaim) [inline src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java:3]: // This class represents an error when a product is out of stock.

change to product- all the classes are product and not item pls

File: src/main/java/Domain/Exceptions/PaymentFailedException.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package Domain.Exceptions;
+
+public class PaymentFailedException extends RuntimeException{
```
Comment (inbarbenchaim) [inline src/main/java/Domain/Exceptions/PaymentFailedException.java:3]: // This class represents an error when ...

File: src/main/java/Domain/Exceptions/ShippingFailedException.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package Domain.Exceptions;
+
+public class ShippingFailedException extends RuntimeException{
```
Comment (inbarbenchaim) [inline src/main/java/Domain/Exceptions/ShippingFailedException.java:3]: // This class represents an error when ...

File: src/main/java/Domain/Product.java [hunk 5]
```diff
@@ -29,6 +31,17 @@ public double getPrice() {
         return _price;
     }
 
+    public void purchaseProduct() {
```
Comment (inbarbenchaim) [inline src/main/java/Domain/Product.java:34]: here you have the name product :)

File: src/main/java/Domain/Product.java [hunk 6]
```diff
@@ -29,6 +31,17 @@ public double getPrice() {
...
+        if (_quantity == 0) {
+            throw new ItemOutOfStockExepction("Product is out of stock");
```
Comment (inbarbenchaim) [inline src/main/java/Domain/Product.java:36]: add log info pls to inform the update of the _quantity of the product

File: src/main/java/Domain/ShoppingBasket.java [hunk 7]
```diff
@@ -3,16 +3,19 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import Domain.Exceptions.ItemOutOfStockExepction;
+
```
Comment (inbarbenchaim) [inline src/main/java/Domain/ShoppingBasket.java:7]: remove one space pls

File: src/main/java/Domain/ShoppingBasket.java [hunk 8]
```diff
@@ -3,16 +3,19 @@
...
     private Integer _shopId;
     private List<Product> _productList;
     private double _basketTotalAmount;
 
     // Constructor
     public ShoppingBasket(Integer shopId){
-        this._shopId = shopId;
-        this._productList = new ArrayList<>();
-        this._basketTotalAmount = 0.0;
+        _shopId = shopId;
```
Comment (inbarbenchaim) [inline src/main/java/Domain/ShoppingBasket.java:16]: THANKYOUUUUU

File: src/main/java/Domain/ShoppingBasket.java [hunk 9]
```diff
@@ -33,6 +36,33 @@ public double getShoppingBasketPrice() {
         return _basketTotalAmount;
     }
 
+    /*
+     * Go through the list of products in the basket and purchase them.
+     * If an exception is thrown, cancel the purchase of all the products that were bought.
+     * This function only updates the item's stock.
+     */
+    public void purchaseBasket() {
+        List<Product> boughtProductList = new ArrayList<>();
```
Comment (inbarbenchaim) [inline src/main/java/Domain/ShoppingBasket.java:45]: log info: start the purchase progress for id.... user name..... bla bla

File: src/main/java/Domain/ShoppingBasket.java [hunk 10]
```diff
@@ -33,6 +36,33 @@ public double getShoppingBasketPrice() {
...
+        for (Product product : _productList) {
+            try{
+                product.purchaseProduct();
+                boughtProductList.add(product);
+            }
+            catch (ItemOutOfStockExepction e){
+                for (Product boughtProduct : boughtProductList) {
+                    boughtProduct.cancelPurchase();
+                }
+                throw new ItemOutOfStockExepction("One of the products in the basket is out of stock");
```
Comment (inbarbenchaim) [inline src/main/java/Domain/ShoppingBasket.java:55]: why exception and not worning? do we want the system to stop? or just inform the user on the items he has? if on items left at all in the stock of maybe error

File: src/main/java/Domain/ShoppingCart.java [hunk 11]
```diff
@@ -1,5 +1,54 @@
...
+
+    public ShoppingCart() {
+        _shoppingBaskets = new ArrayList<>();
+    }
+
+    /*
+     * Go thorugh the list of baskets to buy and purchase them.
+     * If an exception is thrown, cancel the purchase of all the baskets that were bought.
+     * This function only updates the item's stock. 
+     */
```
Comment (inbarbenchaim) [inline src/main/java/Domain/ShoppingCart.java:19]: consider my comment in ShoppingBasket also here pls

File: src/main/java/Domain/ShoppingCart.java [hunk 12]
```diff
@@ -1,5 +1,54 @@
...
+    public void purchaseCart(User buyer, List<Integer> busketsToBuy) {
```
Comment (inbarbenchaim) [inline src/main/java/Domain/ShoppingCart.java:20]: where is the connection to payment and shipping? if not implemented yet pls add //TODO: .... for later to do not forget

File: src/main/java/Domain/User.java [hunk 13]
```diff
@@ -1,46 +1,77 @@
...
+import Domain.Exceptions.ShippingFailedException;
+import Domain.ExternalServices.PaymentService.PaymentMethod;
+import Domain.ExternalServices.SupplyService.SupplyMethod;
+
 public class User {
-    private String encoded_password;
-    private String username;
-    private boolean loggedIn;
-    private String email;
+    private String _encoded_password;
```
Comment (inbarbenchaim) [inline src/main/java/Domain/User.java:11]: THANK YOU YOU KING

File: src/main/java/Domain/User.java [hunk 14]
```diff
@@ -1,46 +1,77 @@
...
+    /*
+     * This method is responsible for purchasing the cart.
+     * It first calls the purchaseCart method of the shopping cart which reaponsible for changing the item's stock.
+     * Then it tries to pay and deliver the items.
+     * If the payment or the delivery fails, it cancels the purchase and restock the item.
+     */
+    public void purchaseCart(List<Integer> busketsToBuy, PaymentMethod paymentMethod, SupplyMethod shippingMethod) {
+        _shoppingCart.purchaseCart(this, busketsToBuy);
+        try {
+            paymentMethod.pay();
```
Comment (inbarbenchaim) [inline src/main/java/Domain/User.java:65]: pls add logs like starting purachase progress, and check the payment and so on.

File: src/main/java/Domain/User.java [hunk 15]
```diff
@@ -1,46 +1,77 @@
...
+    }
+
+    /*
+     * This method is responsible for purchasing the cart.
+     * It first calls the purchaseCart method of the shopping cart which reaponsible for changing the item's stock.
+     * Then it tries to pay and deliver the items.
+     * If the payment or the delivery fails, it cancels the purchase and restock the item.
+     */
+    public void purchaseCart(List<Integer> busketsToBuy, PaymentMethod paymentMethod, SupplyMethod shippingMethod) {
+        _shoppingCart.purchaseCart(this, busketsToBuy);
```
Comment (inbarbenchaim) [inline src/main/java/Domain/User.java:63]: also may raise an error

File: src/main/java/Domain/User.java [hunk 16]
```diff
@@ -1,46 +1,77 @@
...
+        try {
+            paymentMethod.pay();
+            shippingMethod.deliver();
+        } catch (PaymentFailedException e) {
+            _shoppingCart.cancelPurchase(this, busketsToBuy);
+            throw new PaymentFailedException("Payment failed");
+        } catch (ShippingFailedException e) {
+            _shoppingCart.cancelPurchase(this, busketsToBuy);
+            throw new ShippingFailedException("Shipping failed");
```
Comment (inbarbenchaim) [inline src/main/java/Domain/User.java:72]: log SERVER also pls

File: src/main/java/Domain/User.java [hunk 17]
```diff
@@ -1,46 +1,77 @@
...
+        _email = email;
+    }
+
+    /*
+     * This method is responsible for purchasing the cart.
+     * It first calls the purchaseCart method of the shopping cart which reaponsible for changing the item's stock.
+     * Then it tries to pay and deliver the items.
+     * If the payment or the delivery fails, it cancels the purchase and restock the item.
+     */
+    public void purchaseCart(List<Integer> busketsToBuy, PaymentMethod paymentMethod, SupplyMethod shippingMethod) {
```
Comment (inbarbenchaim) [inline src/main/java/Domain/User.java:62]: thinking out load- we didnt say we want this in shopping cart / basket classes? the payment and shipping, not in the user. what do you say?

File: src/main/java/ServiceLayer/UserService.java [hunk 18]
```diff
@@ -61,4 +64,22 @@ public Response register(String userName, String password, String email) {
...
     }
+
+    public Response purchaseCart(String token, List<Integer> busketsToBuy, PaymentMethod paymentMethod, SupplyMethod shippingMethod) {
+        Response response = new Response();
+        try {
+            String username = tokenService.getUsernameFromToken(token);
+            if (userController.isUserNameExists(username)) {
+                userController.getUserByUsername(username).purchaseCart(busketsToBuy, paymentMethod, shippingMethod);
+                logger.info("User purchase cart successfully: " + username);
+                response.setReturnValue("Purchase cart Succeed");
```
Comment (inbarbenchaim) [inline src/main/java/ServiceLayer/UserService.java:75]: you want to return the cart no? or the order?

File: src/main/java/ServiceLayer/UserService.java [hunk 19]
```diff
@@ -61,4 +64,22 @@ public Response register(String userName, String password, String email) {
         }
         return response;
     }
+
+    public Response purchaseCart(String token, List<Integer> busketsToBuy, PaymentMethod paymentMethod, SupplyMethod shippingMethod) {
+        Response response = new Response();
+        try {
+            String username = tokenService.getUsernameFromToken(token);
```
Comment (inbarbenchaim) [inline src/main/java/ServiceLayer/UserService.java:71]: notice the conversation in the group whatsup - you can also buy if you are a guest in the system. I think its important of the saving the orders for the history if he loggs in. also you did not mention the saving in history purchase for the user and in the store- you can write //TODO just so we dont forget its suppose to happened also

Comment (inbarbenchaim) [reply src/main/java/ServiceLayer/UserService.java:71]: Metar got the history task. assigned to him

File: src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java (hunk 1 above)
Comment (TalKoren11) [reply src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java:1]: are you talking about the Exceptions? its not external tho, its home made exceptions :)

File: src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java (hunk 2 above)
Comment (TalKoren11) [reply src/main/java/Domain/Exceptions/ItemOutOfStockExepction.java:3]: ok

File: src/main/java/Domain/Exceptions/PaymentFailedException.java (hunk 3 above)
Comment (TalKoren11) [reply src/main/java/Domain/Exceptions/PaymentFailedException.java:3]: Done.

File: src/main/java/Domain/Exceptions/ShippingFailedException.java (hunk 4 above)
Comment (TalKoren11) [reply src/main/java/Domain/Exceptions/ShippingFailedException.java:3]: Done.

File: src/main/java/Domain/Product.java (hunk 5 above)
Comment (TalKoren11) [reply src/main/java/Domain/Product.java:34]: great!

File: src/main/java/Domain/Product.java (hunk 6 above)
Comment (TalKoren11) [reply src/main/java/Domain/Product.java]: Done.

File: src/main/java/Domain/ShoppingBasket.java (hunk 7 above)
Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java:7]: Done.

File: src/main/java/Domain/ShoppingBasket.java (hunk 8 above)
Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java:16]: great!

File: src/main/java/Domain/User.java (hunk 17 above)
Comment (TalKoren11) [reply src/main/java/Domain/User.java:62]: I will exlain in voice call I think it makes more sence here.

File: src/main/java/Domain/User.java (hunk 15 above)
Comment (TalKoren11) [reply src/main/java/Domain/User.java:63]: which we will catch in the service layer.
Do you think I sould catch it here?

File: src/main/java/Domain/ShoppingBasket.java (hunk 10 above)
Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java]: I dont understand

Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java]: I can change it to return boolean and make it so it wont thorw exception but I am handeling the exception in the function that calls this function, I change the signatures to throw exception for now so we wont forget it.

Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java]: I agree its not the best idea but I think its the best for now, I will change it in future if needed.

Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java]: Done.

File: src/main/java/Domain/ShoppingCart.java (hunk 11 above)
Comment (inbarbenchaim) [reply src/main/java/Domain/ShoppingCart.java:22]: no idea, sorry.

File: src/main/java/Domain/ShoppingBasket.java (hunk 9 above)
Comment (TalKoren11) [reply src/main/java/Domain/ShoppingBasket.java:48]: Done.

File: src/main/java/Domain/ShoppingCart.java (hunk 11 above)
Comment (TalKoren11) [reply src/main/java/Domain/ShoppingCart.java:22]: Done.

File: src/main/java/Domain/User.java (hunk 13 above)
Comment (TalKoren11) [reply src/main/java/Domain/User.java:13]: Done.

File: src/main/java/Domain/User.java (hunk 14 above)
Comment (TalKoren11) [reply src/main/java/Domain/User.java:69]: Done.

File: src/main/java/Domain/User.java (hunk 15 above)
Comment (TalKoren11) [reply src/main/java/Domain/User.java:67]: it cannot raise an error, purchaseCart in ShippingCart is not suppose to throw an error.

Comment (TalKoren11) [reply src/main/java/Domain/User.java:67]: Done.

File: src/main/java/Domain/User.java (hunk 16 above)
Comment (TalKoren11) [reply src/main/java/Domain/User.java:78]: Done.

File: src/main/java/ServiceLayer/UserService.java (hunk 18 above)
Comment (TalKoren11) [reply src/main/java/ServiceLayer/UserService.java:75]: It needs to return order.

**Review by inbarbenchaim — APPROVED** [review]

**Pull request merged**